File: streampark-flink/streampark-flink-connector/streampark-flink-connector-jdbc/src/main/scala/org/apache/streampark/flink/connector/jdbc/internal/JdbcSourceFunction.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.jdbc.internal

import org.apache.streampark.common.enums.ApiType
import org.apache.streampark.common.enums.ApiType.ApiType
import org.apache.streampark.common.util.{JdbcUtils, Logger}
import org.apache.streampark.flink.connector.function.{RunningFunction, SQLQueryFunction, SQLResultFunction}
import org.apache.streampark.flink.util.FlinkUtils

import org.apache.flink.api.common.state.ListState
import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.runtime.state.{CheckpointListener, FunctionInitializationContext, FunctionSnapshotContext}
import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction
import org.apache.flink.streaming.api.functions.source.RichSourceFunction
import org.apache.flink.streaming.api.functions.source.SourceFunction.SourceContext

import java.lang
import java.util.Properties

import scala.collection.Map
import scala.util.{Success, Try}

class JdbcSourceFunction[R: TypeInformation](apiType: ApiType = ApiType.scala, jdbc: Properties)
  extends RichSourceFunction[R]
  with CheckpointedFunction
  with CheckpointListener
  with Logger {

  @volatile private[this] var running = true
  private[this] var scalaRunningFunc: Unit => Boolean = (_) => true
  private[this] var javaRunningFunc: RunningFunction = _

  private[this] var scalaSqlFunc: R => String = _
  private[this] var scalaResultFunc: Function[Iterable[Map[String, _]], Iterable[R]] = _
  private[this] var javaSqlFunc: SQLQueryFunction[R] = _
  private[this] var javaResultFunc: SQLResultFunction[R] = _
  @transient private var state: ListState[R] = _
  private val OFFSETS_STATE_NAME: String = "jdbc-source-query-states"
  private[this] var last: R = _

  // for Scala
  def this(
      jdbc: Properties,
      sqlFunc: R => String,
      resultFunc: Iterable[Map[String, _]] => Iterable[R],
      runningFunc: Unit => Boolean) = {

    this(ApiType.scala, jdbc)
    this.scalaSqlFunc = sqlFunc
    this.scalaResultFunc = resultFunc
    this.scalaRunningFunc = if (runningFunc == null) _ => true else runningFunc
  }

  // for JAVA
  def this(
      jdbc: Properties,
      javaSqlFunc: SQLQueryFunction[R],
      javaResultFunc: SQLResultFunction[R],
      runningFunc: RunningFunction) {

    this(ApiType.java, jdbc)
    this.javaSqlFunc = javaSqlFunc
    this.javaResultFunc = javaResultFunc
    this.javaRunningFunc =
      if (runningFunc != null) runningFunc
      else
        new RunningFunction {
          override def running(): lang.Boolean = true
        }
  }

  @throws[Exception]
  override def run(ctx: SourceContext[R]): Unit = {
    while (this.running) {
      apiType match {
        case ApiType.scala =>
          if (scalaRunningFunc()) {
            ctx.getCheckpointLock.synchronized {
              val sql = scalaSqlFunc(last)
              val result: List[Map[String, _]] = JdbcUtils.select(sql)(jdbc)
              scalaResultFunc(result).foreach(
                x => {
                  last = x
                  ctx.collectWithTimestamp(last, System.currentTimeMillis())
                })
            }
          }
        case ApiType.java =>
          if (javaRunningFunc.running()) {
            ctx.getCheckpointLock.synchronized {
              val sql = javaSqlFunc.query(last)
              val result: List[Map[String, _]] = JdbcUtils.select(sql)(jdbc)
              javaResultFunc
                .result(result.map(_))
                .foreach(
                  x => {
                    last = x
                    ctx.collectWithTimestamp(last, System.currentTimeMillis())
                  })
            }
          }
      }
    }
  }

  override def cancel(): Unit = this.running = false

  override def snapshotState(context: FunctionSnapshotContext): Unit = {
    if (running) {
      state.clear()
      if (last != null) {
        state.add(last)
      }
    } else {
      logError("JdbcSource snapshotState called on closed source")
    }
  }

  override def initializeState(context: FunctionInitializationContext): Unit = {
    logInfo("JdbcSource snapshotState initialize")
    state = FlinkUtils.getUnionListState[R](context, OFFSETS_STATE_NAME)
    Try(state.get.head) match {
      case Success(q) => last = q
      case _ =>
    }
  }

  override def notifyCheckpointComplete(checkpointId: Long): Unit = {
    logInfo(s"JdbcSource checkpointComplete: $checkpointId")
  }
}

File: streampark-flink/streampark-flink-connector/streampark-flink-connector-jdbc/src/main/java/org/apache/streampark/flink/connector/jdbc/internal/JdbcSourceFunction.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.jdbc.internal;

import org.apache.streampark.common.util.JdbcUtils;
import org.apache.streampark.flink.connector.function.RunningFunction;
import org.apache.streampark.flink.connector.function.SQLQueryFunction;
import org.apache.streampark.flink.connector.function.SQLResultFunction;
import org.apache.streampark.flink.util.FlinkUtils;

import org.apache.flink.api.common.state.ListState;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.runtime.state.CheckpointListener;
import org.apache.flink.runtime.state.FunctionInitializationContext;
import org.apache.flink.runtime.state.FunctionSnapshotContext;
import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
import org.apache.flink.streaming.api.functions.source.SourceFunction;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Properties;

/** JDBC source function with checkpoint state support. */
public class JdbcSourceFunction<R> extends RichSourceFunction<R>
        implements CheckpointedFunction, CheckpointListener {

    private static final Logger LOG = LoggerFactory.getLogger(JdbcSourceFunction.class);
    private static final String OFFSETS_STATE_NAME = "jdbc-source-query-states";

    private volatile boolean running = true;
    private final Properties jdbc;
    private final SQLQueryFunction<R> sqlFunc;
    private final SQLResultFunction<R> resultFunc;
    private final RunningFunction runningFunc;
    private final TypeInformation<R> typeInfo;

    private transient ListState<R> state;
    private R last;

    public JdbcSourceFunction(
            Properties jdbc,
            SQLQueryFunction<R> sqlFunc,
            SQLResultFunction<R> resultFunc,
            RunningFunction runningFunc,
            TypeInformation<R> typeInfo) {
        this.jdbc = jdbc;
        this.sqlFunc = sqlFunc;
        this.resultFunc = resultFunc;
        this.runningFunc =
                runningFunc != null
                        ? runningFunc
                        : new RunningFunction() {
                            @Override
                            public Boolean running() {
                                return true;
                            }
                        };
        this.typeInfo = typeInfo;
    }

    @Override
    public void run(SourceContext<R> ctx) throws Exception {
        while (running) {
            if (runningFunc.running()) {
                synchronized (ctx.getCheckpointLock()) {
                    String sql = sqlFunc.query(last);
                    List<Map<String, Object>> result = JdbcUtils.select(sql, jdbc);
                    Iterable<Map<String, ?>> rows = (Iterable) result;
                    Iterable<R> records = resultFunc.result(rows);
                    for (R x : records) {
                        last = x;
                        ctx.collectWithTimestamp(last, System.currentTimeMillis());
                    }
                }
            }
        }
    }

    @Override
    public void cancel() {
        running = false;
    }

    @Override
    public void snapshotState(FunctionSnapshotContext context) throws Exception {
        if (running) {
            state.clear();
            if (last != null) {
                state.add(last);
            }
        } else {
            LOG.error("JdbcSource snapshotState called on closed source");
        }
    }

    @Override
    public void initializeState(FunctionInitializationContext context) throws Exception {
        LOG.info("JdbcSource snapshotState initialize");
        @SuppressWarnings("unchecked")
        TypeInformation<R> stateType =
                typeInfo != null ? typeInfo : (TypeInformation<R>) TypeInformation.of(Object.class);
        state = FlinkUtils.getUnionListState(context, OFFSETS_STATE_NAME, stateType);
        Iterator<R> it = state.get().iterator();
        if (it.hasNext()) {
            last = it.next();
        }
    }

    @Override
    public void notifyCheckpointComplete(long checkpointId) {
        LOG.info("JdbcSource checkpointComplete: {}", checkpointId);
    }
}
